cpp/src/arrow/dataset/file_csv.cc

```diff
@@ -39,49 +41,92 @@ namespace dataset {
 using internal::checked_cast;
 using internal::checked_pointer_cast;
 
+Result<std::unordered_set<std::string>> GetColumnNames(
```

Comment:
Would it make sense/be possible to move this somewhere more generally available for CSV reading? @romainfrancois encountered a similar need in https://github.com/apache/arrow/pull/7807#issuecomment-662045381

Reply:
I wrote this as a separate function in part so that it would be easy to extract as a public function if we wanted that. I'd like to leave that for a follow up, though